🚀 2단계 - 깃헙 로그인 구현

xonic789 opened the pull request:
안녕하세요. 전 단계에서 주셨던 피드백 반영했습니다.

## 미션 진행시 고려했던 점
- 확장을 위해서 OAuth2Controller 를 만들었고, Spring configuration processor를 통해서 환경마다 다른 env를 주입할 수 있도록 구현하였습니다. 그래서 다른 OAuth2 벤더들도 쉽게 interface 만 implement 하여 구현하도록 했습니다.


## 참고 사항
- 시간이 없어서 실제 Github를 통해서 요청하고, accessToken을 받아오지 않고 TestController를 이용하여 code를 내려주면 그에 따라 미리 정의된 응답을 내려주도록 구현했습니다.

xonic789 committed:
doc: 1단게 리드미 추가

xonic789 committed:
test: github 로그인 인수테스트 작성

xonic789 committed:
OAuth2 뼈대 코드 작성

xonic789 committed:
OAuth2Client Test 코드 작성

xonic789 committed:
환경에 따라 다른 env를 주입하도록 spring configuration processor 추가

xonic789 committed:
Github 로그인 인수테스트 작성

xonic789 committed:
변수 제거 - 피드백 반영

testrace commented on README.md:
```diff
@@ -1,2 +1,20 @@
\ No newline at end of file
+[ATDD 강의](https://edu.nextstep.camp/c/R89PYi5H) 실습을 위한 지하철 노선도 애플리케이션
+
+## 🚀 1단계 - 즐겨찾기 기능 완성
```
2단계 미션 요구사항도 추가해 주시면 좋을 것 같아요

testrace commented on build.gradle:
```diff
@@ -26,6 +26,8 @@ dependencies {
     // jwt
     implementation 'io.jsonwebtoken:jjwt:0.9.1'
 
+    annotationProcessor "org.springframework.boot:spring-boot-configuration-processor"
```
configuration-processor 👍

testrace commented on src/main/java/nextstep/AppProperties.java:
```diff
@@ -0,0 +1,69 @@
+package nextstep;
+
+import org.springframework.boot.context.properties.ConfigurationProperties;
+import org.springframework.boot.context.properties.ConstructorBinding;
+
+@ConstructorBinding
+@ConfigurationProperties(prefix = "subway.app")
+public class AppProperties {
```
configuration 관련 패키지를 두어 관리하면 어떨까요?

testrace commented on src/main/java/nextstep/member/application/GithubResponses.java:
```diff
@@ -0,0 +1,30 @@
+package nextstep.member.application;
+
+public enum GithubResponses {
+    사용자1("aofijeowifjaoief", "access_token_1", "email1@email.com"),
+    사용자2("fau3nfin93dmn", "access_token_2", "email2@email.com"),
+    사용자3("afnm93fmdodf", "access_token_3", "email3@email.com"),
+    사용자4("fm04fndkaladmd", "access_token_4", "email4@email.com");
```
GithubResponses는 테스트용 객체로 생각되는데요
test 패키지로 이동하면 어떨까요?

testrace commented on src/main/java/nextstep/member/application/OAuth2Client.java:
```diff
@@ -0,0 +1,6 @@
+package nextstep.member.application;
+
+public interface OAuth2Client {
+
+    String requestAccessToken(String code);
```
client interface 💯

testrace commented on src/main/java/nextstep/member/application/dto/GithubAccessTokenResponse.java:
```diff
@@ -0,0 +1,42 @@
+package nextstep.member.application.dto;
+
+public class GithubAccessTokenResponse {
+
+    private String accessToken;
+    private String scope;
+    private String tokenType;
```
미션에선 scope, type에 대해선 언급이 없는데 추가해주셨네요 👍 
OAuth2를 적용해 보신 적이 있으신 것 같아요 😃

testrace commented on src/main/java/nextstep/member/application/GithubLoginService.java:
```diff
@@ -0,0 +1,23 @@
+    }
+
+
+    public TokenResponse login(OAuth2Request oAuth2Request) {
+        String accessToken = githubClient.requestAccessToken(oAuth2Request.getCode());
+        return new TokenResponse(accessToken);
+    }
```
>요구사항
가입이 되어있지 않은 경우 회원 가입으로 진행 후 토큰 발행

이 부분은 적용이 안되어 있는 것 같아요 😢

testrace commented on src/main/java/nextstep/test/TestController.java:
```diff
@@ -0,0 +1,47 @@
+    public ResponseEntity<GithubAccessTokenResponse> accessToken(
+        @RequestBody GithubAccessTokenRequest tokenRequest) {
+        String accessToken = "";
+        switch (tokenRequest.getCode()) {
+            case "aofijeowifjaoief":
+                accessToken = GithubResponses.사용자1.getAccessToken();
+                break;
+            case "fau3nfin93dmn":
+                accessToken = GithubResponses.사용자2.getAccessToken();
+                break;
+            case "afnm93fmdodf":
+                accessToken = GithubResponses.사용자3.getAccessToken();
+                break;
+            case "fm04fndkaladmd":
+                accessToken = GithubResponses.사용자4.getAccessToken();
+                break;
+        }
```
GithubResponses 객체에게 메시지를 보내서 accessToken을 조회하면 어떨까요?
```java
accessToken = GithubResponses.from(code);
```

testrace commented on src/main/java/nextstep/test/TestController.java:
```diff
@@ -0,0 +1,47 @@
+import org.springframework.web.bind.annotation.RequestHeader;
+import org.springframework.web.bind.annotation.RestController;
+
+@RestController
+public class TestController {
```
테스트용 컨트롤러도 test 패키지로 이동하면 어떨까요?

testrace commented on src/main/resources/application-env.yml:
```diff
@@ -0,0 +1,11 @@
+spring:
+  config:
+    activate:
+      on-profile: test
+subway:
+  app:
+    github:
+      client_id: 'client_id'
+      client_secret: 'client_secret'
+      access_token_uri: http://localhost:8080/github/login/oauth/access_token
```
운영환경에서도 localhost가 적용될 것 같은데요
테스트용 properties를 분리해 보셔도 좋을 것 같아요
`application-test.yml`

testrace commented on src/test/java/nextstep/member/unit/GithubClientTest.java:
```diff
@@ -0,0 +1,32 @@
+
+
+    @Test
+    void requestAccessToken() {
```
테스트에 실패하네요 😢 
<img width="1111" alt="image" src="https://github.com/next-step/atdd-subway-favorite/assets/8064743/7b53d68f-bd2c-48e9-ac98-3c32b1df8bee">

testrace commented on src/test/java/nextstep/member/acceptance/AuthAcceptanceTest.java:
```diff
@@ -35,21 +37,39 @@ void bearerAuth() {
         assertThat(response2.jsonPath().getString("email")).isEqualTo(EMAIL);
     }
+
+    /**
+     * Given code 가 주어진다.
+     * When github 로그인을 시도한다.
+     * Then 회원가입이 완료되고, 로그인이 된다.
+     */
```
이미 github으로 회원가입한 회원으로 로그인하는 경우도 테스트하면 어떨까요?

testrace reviewed (changes requested):
안녕하세요 승우님 😃 
2단계도 잘해주셨네요 👍 
몇 가지 코멘트 남겨두었는데 확인해 주시고 다시 리뷰 요청 부탁드려요

xonic789 committed:
GithubResponses 패키지 변경

xonic789 committed:
prod 적용시 예제

xonic789 committed:
fix: fix test code

xonic789 committed:
이미 깃허브로 로그인 한 사람도 깃허브 로그인 요청이 성공한다.
- AppProperties 패키지 경로 변경
- OAuthClient 인터페이스에 requestUserInfo 추상 메소드 추가 및 구현
- 깃허브 로그인 해피케이스 인수 테스트에 회원 가입 되지 않았다는 걸 체크하는 로직 추가

xonic789 commented:
하다보니, [feb4877](https://github.com/next-step/atdd-subway-favorite/pull/645/commits/feb48771929cb7ce703d816af7cc70afb0cae3dc) 커밋이 좀 커졌는데 양해 부탁드립니다 ㅠㅠ
피드백 주신거 대부분 반영했고, 이미 깃허브로 로그인 한 유저에 대한 해피케이스도 추가하였습니다.

꼼곰히 봐주셔서 감사드립니다 :)

testrace commented on src/main/java/nextstep/member/application/TokenService.java:
```diff
@@ -17,10 +18,9 @@ public TokenService(MemberService memberService, JwtTokenProvider jwtTokenProvid
     public TokenResponse createToken(String email, String password) {
         Member member = memberService.findMemberByEmail(email);
-        if (!member.getPassword().equals(password)) {
+        if (!member.checkPassword(password)) {
```
메세지를 보내도록 변경해 주셨네요 💯

testrace commented on src/main/java/nextstep/test/GithubResponses.java:
```diff
@@ -0,0 +1,65 @@
+package nextstep.test;
+
+public enum GithubResponses {
```
테스트 환경에서만 사용되는 클래스이니 src/main 패키지가 아닌 src/test 패키지에 위치하도록 변경하면 어떨까요?
TestController 마찬가지입니다

testrace commented on src/main/java/nextstep/test/GithubResponses.java:
```diff
@@ -0,0 +1,65 @@
+    }
+
+    public static GithubResponses from(String code) {
+        switch (code) {
+            case "aofijeowifjaoief":
+                return GithubResponses.사용자1;
+            case "fau3nfin93dmn":
+                return GithubResponses.사용자2;
+            case "afnm93fmdodf":
+                return GithubResponses.사용자3;
+            case "fm04fndkaladmd":
+                return GithubResponses.사용자4;
+            case "code":
+                return GithubResponses.이미_가입된_사용자;
+            default:
+                throw new IllegalArgumentException("존재하지 않는 코드입니다.");
```
기능이 많아져서 테스트용 사용자가 10명, 20명으로 늘어나게 된다면 코드 변경량이 많을 것 같은데요
stream을 활용해 switch문을 제거하면 어떨까요?

testrace commented on src/main/resources/application-env.yml:
```diff
@@ -8,4 +8,17 @@ subway:
+spring:
+  config:
+    activate:
+      on-profile: prod
```
프로파일을 추가해주셨군요 👍

testrace commented on src/test/java/nextstep/member/acceptance/AuthAcceptanceTest.java:
```diff
@@ -36,40 +34,60 @@ void bearerAuth() {
+
+    }
+
+    /**
+     * Given 이미 깃허브로 로그인 한 사용자가 주어진다.
+     * When github 로그인을 시도한다.
+     * Then Github 로 이미 가입된 회원이라도 로그인이 성공한다.
+     */
```
인수 조건 👍

testrace reviewed (approved):
안녕하세요 승우님 😃 
피드백 반영 잘해주셨네요 👍 
소소하게 코멘트 남겨두었는데 확인해 주시고 다음 단계 진행해 주세요!
